### PR TITLE
Create convex hull around points

### PR DESCRIPTION
**Pull request by kaushikrw**
## Description
PR to add a new Kotlin sample _"Create convex hull around points"_ in `Analysis` category.

## Links and Data
Sample Epic: `runtime/kotlin/issues/1805`

## What To Review
-  Review the code to make sure it is easy to follow like other samples on Android
- `README.md` and `README.metadata.json` files

## How to Test
- Run the sample on the sample viewer or the repo.
- Tap on the map to add points. Click the "Create Convex Hull" button to generate the convex hull of those points. Click the "Reset" button to start over.

### CODE REVIEW

File: create-convex-hull-around-points/src/main/java/com/esri/arcgismaps/sample/createconvexhullaroundpoints/MainActivity.kt [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+private fun Button.enable() {
+    if (!isEnabled) isEnabled = true
+}
+
+/**
+ * Simple extension function to disable a button, if not already disabled
+ */
+private fun Button.disable() {
+    if (isEnabled) isEnabled = false
+}
```
[the remark this answers is not in the document]
Comment (kaushikrw) [reply create-convex-hull-around-points/src/main/java/com/esri/arcgismaps/sample/createconvexhullaroundpoints/MainActivity.kt:226]: The idea was to not call the`button.isEnabled` setter if it's already set to the desired value, to avoid a re-draw and also for a little bit of cleanliness.

But looks like the setter actually checks for this 
`if (enabled == isEnabled()) return;`

Removing it for redundancy.

File: create-convex-hull-around-points/src/main/java/com/esri/arcgismaps/sample/createconvexhullaroundpoints/MainActivity.kt [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+
+    /**
+     * Creates and draws a convex hull graphic on the map using [pointGeometry] points
+     */
+    private fun createConvexHull(pointGeometry: Geometry) {
+        // normalize the geometry for panning beyond the meridian
+        // and check to proceed if the resulting geometry is not null
+        GeometryEngine.normalizeCentralMeridian(pointGeometry)?.let { normalizedPointGeometry ->
+            // create a convex hull from the points and proceed if it's not null
+            GeometryEngine.convexHull(normalizedPointGeometry)?.let { convexHullGeometry ->
```
[the remark this answers is not in the document]
Comment (kaushikrw) [reply create-convex-hull-around-points/src/main/java/com/esri/arcgismaps/sample/createconvexhullaroundpoints/MainActivity.kt]: Changed it to a run with the `?:` operator to also show an error message.

File: create-convex-hull-around-points/src/main/java/com/esri/arcgismaps/sample/createconvexhullaroundpoints/MainActivity.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (fpearson1111) [reply create-convex-hull-around-points/src/main/java/com/esri/arcgismaps/sample/createconvexhullaroundpoints/MainActivity.kt]: Oh cool, I didn't know about the re-draw thing, and interesting to see that the setter already considers that.

File: create-convex-hull-around-points/src/main/java/com/esri/arcgismaps/sample/createconvexhullaroundpoints/MainActivity.kt [hunk 3]
```diff
@@ -143,45 +142,49 @@ class MainActivity : AppCompatActivity() {
...
-                // disable the create button until new input points are created
-                createButton.disable()
-            } ?: showError("Error creating convex hull")
+        val normalizedPointGeometry = GeometryEngine.normalizeCentralMeridian(pointGeometry)
+        // check to proceed if the resulting geometry is not null
+        normalizedPointGeometry ?: run {
+            // if it is null, show the error and return
+            showError("Error normalizing point geometry")
+            return
         }
```
Comment (fpearson1111) [inline create-convex-hull-around-points/src/main/java/com/esri/arcgismaps/sample/createconvexhullaroundpoints/MainActivity.kt:160]: I would say if the null branch has multiple statements then it's better to just use an if statement
```kotlin
if (normalizedPointGeometry == null) { // ...
```
But both ways are fine, so I'm happy to approve

**Review by fpearson1111 — APPROVED** [review]
Looks good :+1:

File: create-convex-hull-around-points/src/main/java/com/esri/arcgismaps/sample/createconvexhullaroundpoints/MainActivity.kt [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+        // normalize the geometry for panning beyond the meridian
+        val normalizedPointGeometry = GeometryEngine.normalizeCentralMeridian(pointGeometry)
+        // check to proceed if the resulting geometry is not null
+        if (normalizedPointGeometry == null) {
+            // if it is null, show the error and return
+            showError("Error normalizing point geometry")
+            return
+        }
+        // create a convex hull from the points and proceed if it's not null
+        GeometryEngine.convexHull(normalizedPointGeometry)?.let { convexHullGeometry ->
```
Comment (shubham7109) [inline create-convex-hull-around-points/src/main/java/com/esri/arcgismaps/sample/createconvexhullaroundpoints/MainActivity.kt:162]:
```suggestion
        // create a convex hull from the points and proceed if it's not null
        val convexHullGeometry = GeometryEngine.convexHull(normalizedPointGeometry)
            ?: return showError("Error creating convex hull")
```
The elvis operator was making it a little hard to follow. Maybe just get the value and return if null right then.

**Review by shubham7109 — CHANGES_REQUESTED** [review]
Minor comment. But sample looks good 👍

File: create-convex-hull-around-points/src/main/java/com/esri/arcgismaps/sample/createconvexhullaroundpoints/MainActivity.kt (hunk 4 above)
Comment (kaushikrw) [reply create-convex-hull-around-points/src/main/java/com/esri/arcgismaps/sample/createconvexhullaroundpoints/MainActivity.kt:162]: I thought `?.<scope function> {} ?: ` was a generally accepted pattern? I did see some other samples using a similar pattern.

Comment (shubham7109) [reply create-convex-hull-around-points/src/main/java/com/esri/arcgismaps/sample/createconvexhullaroundpoints/MainActivity.kt:162]: Yes, that's true, but it would be best to use it for single-assignment operations. Comment by Changan regarding the Elvis operator: [Link](https://github.com/Esri/arcgis-maps-sdk-kotlin-samples/pull/62#discussion_r1105144591)

Comment (kaushikrw) [reply create-convex-hull-around-points/src/main/java/com/esri/arcgismaps/sample/createconvexhullaroundpoints/MainActivity.kt]: I think in that PR, Changan was referring to the elvis operator branching into a `when` statement instead of handling a nullability clause. 

I've refactored it to a traditional `if` clause.

File: create-convex-hull-around-points/src/main/java/com/esri/arcgismaps/sample/createconvexhullaroundpoints/MainActivity.kt [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+import androidx.databinding.DataBindingUtil
+import androidx.lifecycle.lifecycleScope
+import com.arcgismaps.ApiKey
+import com.arcgismaps.ArcGISEnvironment
+import com.arcgismaps.Color
+import com.arcgismaps.geometry.*
+import com.arcgismaps.mapping.ArcGISMap
+import com.arcgismaps.mapping.BasemapStyle
+import com.arcgismaps.mapping.Viewpoint
+import com.arcgismaps.mapping.symbology.*
```
Comment (shubham7109) [inline create-convex-hull-around-points/src/main/java/com/esri/arcgismaps/sample/createconvexhullaroundpoints/MainActivity.kt:32]: We try to use only API-specific imports instead of importing all using `import com.arcgismaps.geometry.*`. 
The medatata script and the sample viewer would use these imports for the discoverability of samples.

File: create-convex-hull-around-points/src/main/java/com/esri/arcgismaps/sample/createconvexhullaroundpoints/MainActivity.kt [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+            this.map = map
+            // add the graphics overlay to the mapview
+            graphicsOverlays.add(graphicsOverlay)
+        }
+
+        lifecycleScope.launch {
+            // if the map load fails show the error and return
+            map.load().onFailure {
+                showError("Error loading map")
+                return@launch
```
Comment (shubham7109) [inline create-convex-hull-around-points/src/main/java/com/esri/arcgismaps/sample/createconvexhullaroundpoints/MainActivity.kt:112]:
```suggestion
                return@launch showError("Error loading map")
```
Could be condensed here.

File: create-convex-hull-around-points/src/main/java/com/esri/arcgismaps/sample/createconvexhullaroundpoints/MainActivity.kt [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+     */
+    private fun createConvexHull(pointGeometry: Geometry) {
+        // normalize the geometry for panning beyond the meridian
+        val normalizedPointGeometry = GeometryEngine.normalizeCentralMeridian(pointGeometry)
+        // check to proceed if the resulting geometry is not null
+        if (normalizedPointGeometry == null) {
+            // if it is null, show the error and return
+            showError("Error normalizing point geometry")
+            return
+        }
```
Comment (shubham7109) [inline create-convex-hull-around-points/src/main/java/com/esri/arcgismaps/sample/createconvexhullaroundpoints/MainActivity.kt:160]:
```suggestion
        // normalize the geometry for panning beyond the meridian
        val normalizedPointGeometry = GeometryEngine.normalizeCentralMeridian(pointGeometry)
            ?: return showError("Error normalizing point geometry")
```
Here would be an ideal example to use the `?:` elvis operator for a single line `val` assignment.

File: create-convex-hull-around-points/src/main/java/com/esri/arcgismaps/sample/createconvexhullaroundpoints/MainActivity.kt [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+        // create a convex hull from the points and proceed if it's not null
+        val convexHullGeometry = GeometryEngine.convexHull(normalizedPointGeometry)
+        if (convexHullGeometry == null) {
+            // if the convex hull is null, show the error and return
+            showError("Error creating convex hull")
+            return
+        }
```
Comment (shubham7109) [inline create-convex-hull-around-points/src/main/java/com/esri/arcgismaps/sample/createconvexhullaroundpoints/MainActivity.kt:167]:
```suggestion
        // create a convex hull from the points and proceed if it's not null
        val convexHullGeometry = GeometryEngine.convexHull(normalizedPointGeometry)
            ?: return showError("Error creating convex hull")
```

Same as above.

File: create-convex-hull-around-points/src/main/res/layout/activity_main.xml [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+        <com.google.android.material.button.MaterialButton
+            android:id="@+id/resetButton"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:enabled="false"
+            android:text="@string/reset"
+            app:layout_constraintBottom_toTopOf="@id/createButton"
+            app:layout_constraintEnd_toEndOf="@id/createButton" />
+
```
Comment (shubham7109) [inline create-convex-hull-around-points/src/main/res/layout/activity_main.xml:37]:
```suggestion
```
Extra line space

File: create-convex-hull-around-points/src/main/res/layout/activity_main.xml [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+            android:textAllCaps="false"
+            android:textSize="14sp"
+            app:layout_constraintBottom_toBottomOf="parent"
+            app:layout_constraintEnd_toEndOf="parent" />
+
+        <com.google.android.material.button.MaterialButton
+            android:id="@+id/resetButton"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:enabled="false"
```
Comment (shubham7109) [inline create-convex-hull-around-points/src/main/res/layout/activity_main.xml:33]:
```suggestion
            android:enabled="false"
            android:textAllCaps="false"
```

Since one of the buttons have `textAllCaps` to false, would be good for consistency.

**Review by shubham7109 — CHANGES_REQUESTED** [review]
Found a couple of things to address, then should be good to approve.

**Review by shubham7109 — APPROVED** [review]
Looks good 👍
